### ISSUE TITLE
Passwords stored in cleartext for webdav in MongoDB

### ISSUE THREAD

**Issue by seppultra** [RocketChat/Rocket.Chat#19258]
### Description:

The WebDAV accounts for every user are stored in the mongoDB. Here the passwords are stored in cleartext. Actually it's a serious security issue if you're using Rocketchat in a company.

### Steps to reproduce:

1. Add a WebDAV Account in your Rocket.Chat Client
2. Go to your MongoDB Server/Container and open mongodb shell:
2.1 `use rocketchat`
2.2 `db.rocketchat_webdav_accounts.find()`

### Expected behavior:

Some hashed passwords

### Actual behavior:

Passwords are stored in cleartext and for everyone with db access visible

### Server Setup Information:

- Version of Rocket.Chat Server: 3.7.0
- Deployment Method: Docker
- MongoDB Version: 4.0

### Client Setup Information

- Desktop App or Browser Version: 3.0.3
- Operating System: Every OS

**Comment by fdellwing** [comment]
RC uses this password to connect to some WebDAV resource, right? So it needs access to the plaintext password for this. Any encryption on this value would therefore be snake oil, because RC is open source and the decrypt method would be publicly accessible.

Thinking about this, it comes to me that it may be possible to encrypt it with a/the user password. To decrypt it, the user would than have to supply a/his password. I don't know if this is feasible to use in this case.

**Comment by seppultra** [comment]
Well i'm not a security expert but does it really needs the plaintext password? When the user entered the password and it will stored hashed in the database. RC could just compare the local hashed password with the WebDAV hashed password and it works. Not?
Another solution could be to store it locally and use Cookies or something similar. 
How does Rocket.Chat itself knows it is still logged on? Maybe the WebDAV Login should be stored at the same way instead of globally in one database.

**Comment by Exordian** [comment]
> RC uses this password to connect to some WebDAV resource, right? So it needs access to the plaintext password for this. Any encryption on this value would therefore be snake oil, because RC is open source and the decrypt method would be publicly accessible.

I don't think so. It's quite common that the database is backuped and/or operated by different persons/org units. Injecting an encryption key into the application and encrypt sensitive information within the database is quite standard. Yet the encryption key shouldn't be stored in the database itself.

> Thinking about this, it comes to me that it may be possible to encrypt it with a/the user password. To decrypt it, the user would than have to supply a/his password. I don't know if this is feasible to use in this case.

We're using SAML for authentication. For federated authentication this approach won't work.
